File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/compbot/LocalizerComp.java
package org.firstinspires.ftc.teamcode.hardware.compbot;

import androidx.annotation.NonNull;

import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.roadrunner.geometry.Pose2d;
import com.acmerobotics.roadrunner.localization.ThreeTrackingWheelLocalizer;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.firstinspires.ftc.teamcode.util.Encoder;

import java.util.Arrays;
import java.util.List;

/*
 * Sample tracking wheel localizer implementation assuming the standard configuration:
 *
 *    /--------------\
 *    |     ____     |
 *    |     ----     |
 *    | ||        || |
 *    | ||        || |
 *    |              |
 *    |              |
 *    \--------------/
 *
 */
@Config
public class LocalizerComp extends ThreeTrackingWheelLocalizer {
    public static double ticksPerRev = 2400;
    public static double wheelRadius = 1.5; // in
    public static double gearRatio = 1; // output (wheel) speed / input (encoder) speed

    public static double LATERAL_DISTANCE = 17.3; // in; distance between the left and right wheels
    public static double FORWARD_OFFSET = -3.6; // in; offset of the lateral wheel

    public static boolean LEFT_REVERSED = true;
    public static boolean RIGHT_REVERSED = true;
    public static boolean FRONT_REVERSED = true;

    public static double X_MULTIPLIER = 1.04;
    public static double Y_MULTIPLIER = 1.04;

    private Encoder leftEncoder, rightEncoder, frontEncoder;

    public LocalizerComp(HardwareMap hardwareMap) {
        super(Arrays.asList(
                new Pose2d(0, LATERAL_DISTANCE / 2, 0), // left
                new Pose2d(0, -LATERAL_DISTANCE / 2, 0), // right
                new Pose2d(FORWARD_OFFSET, 0, Math.toRadians(90)) // front
        ));

        leftEncoder = new Encoder(hardwareMap.get(DcMotorEx.class, "LF"));
        rightEncoder = new Encoder(hardwareMap.get(DcMotorEx.class, "RF"));
        frontEncoder = new Encoder(hardwareMap.get(DcMotorEx.class, "LB"));

        // FINISHED: reverse any encoders using Encoder.setDirection(Encoder.Direction.REVERSE)
        if(RIGHT_REVERSED) rightEncoder.setDirection(Encoder.Direction.REVERSE);
        if(LEFT_REVERSED) leftEncoder.setDirection(Encoder.Direction.REVERSE);
        if(FRONT_REVERSED) frontEncoder.setDirection(Encoder.Direction.REVERSE);
    }

    public static double encoderTicksToInches(double ticks) {
        return wheelRadius * 2 * Math.PI * gearRatio * ticks / ticksPerRev;
    }

    @NonNull
    @Override
    public List<Double> getWheelPositions() {
        return Arrays.asList(
                encoderTicksToInches(leftEncoder.getCurrentPosition()) * X_MULTIPLIER,
                encoderTicksToInches(rightEncoder.getCurrentPosition()) * X_MULTIPLIER,
                encoderTicksToInches(frontEncoder.getCurrentPosition()) * Y_MULTIPLIER
        );
    }

    @NonNull
    @Override
    public List<Double> getWheelVelocities() {
        // FINISHED: If your encoder velocity can exceed 32767 counts / second (such as the REV Through Bore and other
        //  competing magnetic encoders), change Encoder.getRawVelocity() to Encoder.getCorrectedVelocity() to enable a
        //  compensation method

        return Arrays.asList(
                encoderTicksToInches(leftEncoder.getRawVelocity()) * X_MULTIPLIER,
                encoderTicksToInches(rightEncoder.getRawVelocity()) * X_MULTIPLIER,
                encoderTicksToInches(frontEncoder.getRawVelocity()) * Y_MULTIPLIER
        );
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/compbot/LocalizerComp.kt
package org.firstinspires.ftc.teamcode.hardware.compbot

import com.acmerobotics.dashboard.config.Config
import com.acmerobotics.roadrunner.geometry.Pose2d
import com.acmerobotics.roadrunner.localization.ThreeTrackingWheelLocalizer
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.teamcode.util.Encoder
import java.util.*

/*
* Sample tracking wheel localizer implementation assuming the standard configuration:
*
*    /--------------\
*    |     ____     |
*    |     ----     |
*    | ||        || |
*    | ||        || |
*    |              |
*    |              |
*    \--------------/
*
*/
@Config
class LocalizerComp(hardwareMap: HardwareMap) : ThreeTrackingWheelLocalizer(listOf(
        Pose2d(0.0, LATERAL_DISTANCE / 2, 0.0),  // left
        Pose2d(0.0, -LATERAL_DISTANCE / 2, 0.0),  // right
        Pose2d(FORWARD_OFFSET, 0.0, Math.toRadians(90.0)) // front
)) {
    private val leftEncoder: Encoder = Encoder(hardwareMap.get(DcMotorEx::class.java, "LF"))
    private val rightEncoder: Encoder = Encoder(hardwareMap.get(DcMotorEx::class.java, "RF"))
    private val frontEncoder: Encoder = Encoder(hardwareMap.get(DcMotorEx::class.java, "LB"))

    init {
        // FINISHED: reverse any encoders using Encoder.setDirection(Encoder.Direction.REVERSE)
        if (RIGHT_REVERSED) rightEncoder.direction = Encoder.Direction.REVERSE
        if (LEFT_REVERSED) leftEncoder.direction = Encoder.Direction.REVERSE
        if (FRONT_REVERSED) frontEncoder.direction = Encoder.Direction.REVERSE
    }

    override fun getWheelPositions(): List<Double> {
        return listOf(
                encoderTicksToInches(leftEncoder.currentPosition.toDouble()) * X_MULTIPLIER,
                encoderTicksToInches(rightEncoder.currentPosition.toDouble()) * X_MULTIPLIER,
                encoderTicksToInches(frontEncoder.currentPosition.toDouble()) * Y_MULTIPLIER
        )
    }

    override fun getWheelVelocities(): List<Double> {
        // FINISHED: If your encoder velocity can exceed 32767 counts / second (such as the REV Through Bore and other
        //  competing magnetic encoders), change Encoder.getRawVelocity() to Encoder.getCorrectedVelocity() to enable a
        //  compensation method
        return listOf(
                encoderTicksToInches(leftEncoder.rawVelocity) * X_MULTIPLIER,
                encoderTicksToInches(rightEncoder.rawVelocity) * X_MULTIPLIER,
                encoderTicksToInches(frontEncoder.rawVelocity) * Y_MULTIPLIER
        )
    }

    companion object {
        @JvmField
        var ticksPerRev = 2400.0
        @JvmField
        var wheelRadius = 1.5 // in
        @JvmField
        var gearRatio = 1.0 // output (wheel) speed / input (encoder) speed
        @JvmField
        var LATERAL_DISTANCE = 17.3 // in; distance between the left and right wheels
        @JvmField
        var FORWARD_OFFSET = -3.6 // in; offset of the lateral wheel
        @JvmField
        var LEFT_REVERSED = true
        @JvmField
        var RIGHT_REVERSED = true
        @JvmField
        var FRONT_REVERSED = true
        @JvmField
        var X_MULTIPLIER = 1.04
        @JvmField
        var Y_MULTIPLIER = 1.04

        fun encoderTicksToInches(ticks: Double): Double {
            return wheelRadius * 2 * Math.PI * gearRatio * ticks / ticksPerRev
        }
    }
}